ntoskrnl/config/cmapi.c

```diff
@@ -2340,9 +2340,10 @@ CmUnloadKey(IN PCM_KEY_CONTROL_BLOCK Kcb,
 ULONG
 NTAPI
 CmpEnumerateOpenSubKeys(
-    IN PCM_KEY_CONTROL_BLOCK RootKcb,
-    IN BOOLEAN RemoveEmptyCacheEntries,
-    IN BOOLEAN DereferenceOpenedEntries)
+    _In_ PCM_KEY_CONTROL_BLOCK RootKcb,
+    _In_ BOOLEAN LockHeldExclusively,
```

Comment:
Why do you need to add this?

Reply:
>Did you add this variable for the case where the registry is only locked shared?

Yes indeed. The registry is merely locked in shared mode [here](https://github.com/reactos/reactos/blob/master/ntoskrnl/config/ntapi.c#L1906) (when a normal hive unloading request is invoked). Only both the parent and children KCBs (or just the actual target KCB is locked if it has no parent) are locked. This leaves the rest of the Registry infrastructure such as the cache table up to anybody to access.

As a result the cached KCBs of subkeys of the main target KCB (the root one of the hive to name it) are unprotected and could be touched by anybody concurrently.  Removing such a cached KCB without a locking leads to a bunch of asserts in several places of the registry code, for a good reason.

Therefore having this parameter helps us understanding if the registry is entirely locked down exclusively or not. In the former scenario we do not need to lock the individual cached KCB as it's overkill (plus it won't make much sense). When the registry is just locked shared, we gotta lock up the cached KCB in our own.

Reply:
ok this makes sense